### ISSUE TITLE
Flexible rabit function call.

### ISSUE THREAD

**Issue by trivialfis** [dmlc/xgboost#5012]
In https://github.com/dmlc/xgboost/pull/4990#issuecomment-549637875 ,  I felt that this is a serious issue as running something more than once is expected.

@chenqin feel free to reach out if you need any help. ;-)

@CodingCat @hcho3

**Comment by chenqin** [comment]
This topic were covered in previous PR. In nutshell, it's hard to reason which allreduce results failed worker suppose to recovery from without some kind of signature. 

- In one side, raibt implemenation from tianqi etc rely on sequence number per iteration help with failure recovery
- In other side, because bootstrap time recovered worker we often don't have same sequence as first time start workers, we switch to use function call signature to do matching

So we can either allow user do same allreduce call before loadcheckpoint multiple times assuming user wouldn't mess up with order (with help from signature to skip results during scanning) or we offload unique signature duty to users yet offer out of order fetch and recovery primitives.

**Comment by trivialfis** [comment]
@chenqin I will revisit the previous discussion.  Thanks for the clarification.

**Comment by CodingCat** [comment]
if single point recover is immature, can we disable it instead of blocking everything?

**Comment by trivialfis** [comment]
@CodingCat If the PR is reverted I will unblock this issue. I will leave it for @chenqin to decide whether it's infeasible to guarantee all calls to rabit won't trigger this error.

**Comment by CodingCat** [comment]
I think this is related my previous comments on https://github.com/dmlc/rabit/pull/98#discussion_r316969897 ?

**Comment by trivialfis** [comment]
Or at least there should be a canonical way to use it "correctly" and the behaviour being testable.

**Comment by trivialfis** [comment]
@CodingCat Yup.

**Comment by chenqin** [comment]
First of all, new code path need to opt in so there is no blocking whatsoever. The reason why you have seen test failure because new pr introduced duplicated allreduce call why is not support by either old code path nor new code path. The difference was I introduced a Jenkins tests which we disabled to unblock new pr get in. 

Moving forward, I plan to put better debugging message to rabit user so they have better pictures what is supported vs not.

**Comment by trivialfis** [comment]
@chenqin Could you help me understand why it's a duplicated call, and how do I re-enable that test?

**Comment by chenqin** [comment]
> @chenqin Could you help me understand why it's a duplicated call, and how do I re-enable that test?

explain duplication:

void foo(){
   rabit::allreduce(.....)
}

rabit::init()
//duplicated as we use caller name and file line#
foo()
foo()
rabit::loadcheckpoint()
for(loop){
//not duplicated as we use seq number
foo()
foo()
rabit::checkpoint()
}
rabit::shutdown()

how to enable:

void foo(string unique= ""){
   rabit::allreduce(...., unique)
}

rabit::init()
//no more collision
foo("1")
foo("2")
rabit::loadcheckpoint()
for(loop){
//doesn't matter
foo()
foo()
rabit::checkpoint()
}
rabit::shutdown()

**Comment by trivialfis** [comment]
Okay.  Thanks!  Will create a PR trying to workaround it.

**Comment by chenqin** [comment]
Hold on this for a min, I am considering reduce this kind of burden. Let me try on rabit side first

**Comment by trivialfis** [comment]
@chenqin Any update?  I'm thinking about bringing back your original `is_boostrap` option.  Also I don't think there's any need for recovery during construction of DMatrix as it happens before generating model (training).  Do you think there might an easy way to disable recovery at such cases?

**Comment by chenqin** [comment]
> @chenqin Any update? I'm thinking about bringing back your original `is_boostrap` option. Also I don't think there's any need for recovery during construction of DMatrix as it happens before generating model (training). Do you think there might an easy way to disable recovery at such cases?

I have been overwhelmed with other things, ETA end of week. bring back is_bootstrap will not fix issue because root cause is DMatrix was constructed in a way that requires all workers talk with each other twice or multiple times. Think of a worker dead and retry, how can it figure out which results it suppose to feedback to user land if there are more than one calls with same signature.

Recovery is opt-in based, so if user don't do anything, it use old code path.

**Comment by trivialfis** [comment]
@chenqin Thanks for the good work.  I'm not urging anything, just trying to know your progress.  Please let me know if I can be of help.  If you feel I'm nitpicking too much due to lack of knowledge, you can reach me offline.
